### PR TITLE
add an API to allow changing password (if the current password is known) while logged into kotsadm

### PR DESCRIPTION
**Pull request by pavansokkenagaraj**
#### What type of PR is this?

type::feature

#### What this PR does / why we need it:

**add an API to allow changing password  (if the current password is known) while logged into kotsadm**

- add a new `PUT /api/v1/password/change` endpoint to change kots shared password
- request takes current and new password, checks if current password matches, current pwd != new pwd and new pwd length greater than 6.
- once the password is created, deletes all the sessions, so pwd change is enforced
- added the delete all sessions to kots cli cmd `kubectl kots reset-password`
- added a new pacakge `pkg/password` which is being used by cli and api for password change code.
- add a extra check on session validation: if password was updated after session was issued, session will be invalidated and try to delete and log out the user with appropriate message.
- add a new `passwordUpdatedAt` key/value to `kotsadm-password` to record when password was changed (addresses backward compatibility for existing secrets/newly created secrets by using secret.CreatedAt )

#### Which issue(s) this PR fixes:
Fixes # [SC-44273](https://app.shortcut.com/replicated/story/44273/method-to-change-password-in-kots-admin-console-ui#activity-46631)

#### Does this PR introduce a user-facing change?

```release-note
* Fixes an issue where `kubectl kots reset-password` will force the password change by deleting all the user sessions.
* Add an API to allow changing password (if the current password is known) while logged into kotsadm
```

#### Does this PR require documentation?

NONE
(since we don't document API changes, password reset documentation will be part of UI changes)

### CODE REVIEW

File: pkg/store/kotsstore/user_store.go [hunk 1]
```diff
@@ -231,3 +232,64 @@ func (s *KOTSStore) flagSuccessfulLoginInDatabase() error {
 
 	return nil
 }
+
+// SetSharedPasswordBcrypt - set the shared password bcrypt hash in the kotsadm secret
+func (s *KOTSStore) SetSharedPasswordBcrypt(bcryptPassword []byte) error {
```
Comment (laverya) [inline pkg/store/kotsstore/user_store.go:237]: is there not already a function for this, called by the CLI?

Can we use that instead?

Further - do we ever actually use a password from sqlite?

File: cmd/kots/cli/reset-password.go [hunk 2]
```diff
@@ -146,3 +151,29 @@ func setKotsadmPassword(password string, namespace string) error {
...
 }
+
+// delete all sessions from the kotsadm-sessions secret
+func deleteAllSessions(namespace string) error {
+	clientset, err := k8sutil.GetClientset()
+	if err != nil {
+		return errors.Wrap(err, "failed to create k8s client")
+	}
+
+	existingSecret, err := clientset.CoreV1().Secrets(namespace).Get(context.TODO(), "kotsadm-sessions", metav1.GetOptions{})
```
Comment (emosbaugh) [inline cmd/kots/cli/reset-password.go:162]: why do we need to get the secret at all here?

File: pkg/handlers/password_change.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		return
+	}
+	if identityConfig.Spec.Enabled && identityConfig.Spec.DisablePasswordAuth {
+		err := errors.New("password authentication disabled")
+		JSON(w, http.StatusForbidden, NewErrorResponse(err))
+		return
+	}
+
+	var passwordChangeRequest PasswordChangeRequest
+	if err := json.NewDecoder(r.Body).Decode(&passwordChangeRequest); err != nil {
```
Comment (emosbaugh) [inline pkg/handlers/password_change.go:42]: can we move lines 41-51 before `identity.GetConfig` as an optimization?

File: pkg/policy/policies.go [hunk 4]
```diff
@@ -57,6 +57,12 @@ var (
 	PrometheussettingsWrite = Must(NewPolicy(ActionWrite, "prometheussettings."))
 )
 
+// Password change
+
+var (
+	PasswordChange = Must(NewPolicy(ActionWrite, "passwordchange."))
```
Comment (emosbaugh) [inline pkg/policy/policies.go:63]: PasswordWrite or PasswordUpdate for consistency?

File: pkg/store/kotsstore/session_store.go [hunk 5]
```diff
@@ -333,3 +333,21 @@ func (s *KOTSStore) DeleteExpiredSessions() error {
 
 	return nil
 }
+
+// DeleteAllSessions - delete all sessions from kotsadm-sessions secret
+func (s *KOTSStore) DeleteAllSessions() error {
+	sessionLock.Lock()
+	defer sessionLock.Unlock()
+
+	secret, err := s.getSessionSecret()
```
Comment (emosbaugh) [inline pkg/store/kotsstore/session_store.go:342]: why do we need to get the secret at all here?

File: pkg/handlers/password_change.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	if err := user.ValidatePasswordInput(passwordChangeRequest.CurrentPassword, passwordChangeRequest.NewPassword); err != nil {
+		JSON(w, http.StatusBadRequest, NewErrorResponse(err))
+		return
+	}
+
+	// change password
+	if err := user.ChangePassword(store.GetStore(), passwordChangeRequest.CurrentPassword, passwordChangeRequest.NewPassword); err != nil {
+		if err == user.ErrInvalidPassword {
```
Comment (emosbaugh) [inline pkg/handlers/password_change.go:55]: should you change this to the following in case a change is made to the called function?

```
if errors.Is(err, user.ErrInvalidPassword) {
```

File: cmd/kots/cli/reset-password.go [hunk 7]
```diff
@@ -146,3 +151,29 @@ func setKotsadmPassword(password string, namespace string) error {
 
 	return nil
 }
+
+// delete all sessions from the kotsadm-sessions secret
+func deleteAllSessions(namespace string) error {
```
Comment (emosbaugh) [inline cmd/kots/cli/reset-password.go:156]: could we also share this code?

**Review by emosbaugh — CHANGES_REQUESTED** [review]
I do wonder if we could share some more code with the cli as well as andrew pointed out

File: cmd/kots/cli/reset-password.go (hunk 2 above)
Comment (pavansokkenagaraj) [reply cmd/kots/cli/reset-password.go:162]: never knew there was a `func DeleteCollection` , should have checked the interface.

File: cmd/kots/cli/reset-password.go [hunk 8]
```diff
@@ -56,6 +56,11 @@ func ResetPasswordCmd() *cobra.Command {
 				return errors.Wrap(err, "failed to set new password")
 			}
 
+			// delete all sessions to force password change
+			if err := deleteAllSessions(namespace); err != nil {
```
Comment (trelore) [inline cmd/kots/cli/reset-password.go:60]: From a security perspective (I also lack full context, so feel free to point that out):

What happens if `deleteAllSessions` errors here?
- The user gets a exit code `0`.
- The user gets a log statement saying it didn't delete all sessions (I can guarantee someone isn't going to read it if they get `0`).
- A compromised session may still exist.

Other possibly alternatives (in no particular order);
- Error here
  - Compromised sessions may still exist
  - Rerunning the command may produce the same error (though hopefully it's a transient error)
- Put in some kind of retry logic (definitely a route I'd avoid unless we have to/have a pattern for it)
  - This way transient errors are hopefully caught
- Some kind of db style transaction and roll back the password
  - Feels very ugly, and hard to implement if there's no precedent for it
- Make a manual delete sessions CLI command to ensure if this step fails, the user can try just the last bit (and error out here to highlight they have work to do)
  - Adds an edge case for users, but allows them to have more granular control over their password management.
 
In summary: Maybe line 61 needs to be an error to bring attention to the user that they _may_ still have something to do

File: cmd/kots/cli/reset-password.go (hunk 2 above)
Comment (trelore) [reply cmd/kots/cli/reset-password.go:162]: Out of curiousity, what happens if we delete the secret, does it expect a secret to exist to append to, or will the first session create the secret?

File: pkg/handlers/session.go [hunk 9]
```diff
@@ -85,6 +85,20 @@ func requireValidSession(kotsStore store.Store, w http.ResponseWriter, r *http.R
...
+	if passwordUpdatedAt.After(sess.IssuedAt) {
+		if err := kotsStore.DeleteSession(sess.ID); err != nil {
+			logger.Error(errors.Wrapf(err, "password was updated after session created. failed to delete invalid session %s", sess.ID))
+		}
+		err := errors.New("password changed, please login again")
+		response := ErrorResponse{Error: err.Error()}
+		JSON(w, http.StatusUnauthorized, response)
+		return nil, err
+	}
+
```
Comment (laverya) [inline pkg/handlers/session.go:101]: isn't this redundant if the sessions are being deleted?

Comment (pavansokkenagaraj) [reply pkg/handlers/session.go]: if delete sessions fail, this would be a good fallback option 
 additionally, addresses @trelore [comment](https://github.com/replicatedhq/kots/pull/2739#discussion_r856173161)

File: cmd/kots/cli/reset-password.go (hunk 2 above)
Comment (pavansokkenagaraj) [reply cmd/kots/cli/reset-password.go]: @trelore if `kotsadm-secret` gets deleted, kots will fallback to environment password value `SHARED_PASSWORD_BCRYPT` 
when someone changes the secret, the new secret get's created

File: cmd/kots/cli/reset-password.go (hunk 8 above)
Comment (pavansokkenagaraj) [reply cmd/kots/cli/reset-password.go]: @trelore I have added a session check, if a session was created before the passwordUpdatedAt, we invalidate session,  try deleting again and log the user out.
it should address the above comment
https://github.com/replicatedhq/kots/pull/2739#discussion_r856450037

File: pkg/store/kotsstore/user_store.go (hunk 1 above)
Comment (pavansokkenagaraj) [reply pkg/store/kotsstore/user_store.go]: I was bit hesitant to refactor existing code, added a new package `password` which is being used by cli as well as api

File: pkg/store/kotsstore/user_store.go [hunk 10]
```diff
@@ -231,3 +231,35 @@ func (s *KOTSStore) flagSuccessfulLoginInDatabase() error {
...
+	var passwordUpdatedAt time.Time
+	updatedAtBytes, ok := passwordSecret.Data["passwordUpdatedAt"]
+	if ok {
+		updatedAt, err := time.Parse(time.RFC3339, string(updatedAtBytes))
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse passwordUpdatedAt")
+		}
+		passwordUpdatedAt = updatedAt
+	} else {
+		// backward compatibility, for older secret/ newly created secert with no passwordUpdatedAt value
```
Comment (laverya) [inline pkg/store/kotsstore/user_store.go:260]:
```suggestion
		// backward compatibility, for older secret/ newly created secret with no passwordUpdatedAt value
```

**Review by laverya — APPROVED** [review]

File: pkg/password/password_test.go [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+			name: "expect no error when existing secret updated successfully",
+			args: args{
+				clientset:      newMockClientForExistingSecretUpdateSuccess(),
+				namespace:      "test",
+				bcryptPassword: []byte("password"),
+			},
+			wantErr: false,
+		},
+		{
+			name: "expect error when existing secret updated successfully and sessions delete err",
```
Comment (laverya) [inline pkg/password/password_test.go:293]:
```suggestion
			name: "expect no error when existing secret updated successfully and sessions delete err",
```

**Review by laverya — DISMISSED** [review]

**Review by emosbaugh — APPROVED** [review]

File: pkg/password/password.go [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+		}
+
+		return errors.Wrap(err, "failed to compare current password")
+	}
+
+	return nil
+}
+
+// ChangePassword - will change the password in the kotsadm secret
+func ChangePassword(clientset *kubernetes.Clientset, namespace string, newPassword string) error {
```
Comment (sgalsaleh) [inline pkg/password/password.go:55]: needs a mutex

File: pkg/password/password.go [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	if newPassword == currentPassword {
+		return ErrNewPasswordShouldBeDifferent
+	}
+	return nil
+}
+
+// ValidateCurrentPassword - will compare the password with the stored password and return an error if they don't match
+func ValidateCurrentPassword(kotsStore store.Store, currentPassword string) error {
```
Comment (sgalsaleh) [inline pkg/password/password.go:37]: needs a mutex

File: pkg/password/password.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+
+	if err := setSharedPasswordBcrypt(clientset, namespace, shaBytes); err != nil {
+		return errors.Wrap(err, "failed to set new shared password bcrypt")
+	}
+
+	return nil
+}
+
+// setSharedPasswordBcrypt - set the shared password bcrypt hash in the kotsadm secret
+func setSharedPasswordBcrypt(clientset kubernetes.Interface, namespace string, bcryptPassword []byte) error {
```
Comment (sgalsaleh) [inline pkg/password/password.go:69]: needs a mutex

File: pkg/password/password.go (hunk 13 above)
Comment (laverya) [reply pkg/password/password.go:41]: I personally looked at this and thought the same thing at first, but this can be called from both the CLI and the server itself so IMO a mutex is unlikely to help prevent races

Can't hurt though

Comment (emosbaugh) [reply pkg/password/password.go:41]: lgtm although its possible to use resource version in k8s object as a mutex

somthing like validateAndSetSharedPassword

Comment (emosbaugh) [reply pkg/password/password.go:41]: lgtm as i dont think a mutex is really needed but maybe there is some way to use resource version mismatch as a mutex?

**Review by sgalsaleh — APPROVED** [review]

**Review by emosbaugh — APPROVED** [review]

File: pkg/password/password.go (hunk 13 above)
Comment (pavansokkenagaraj) [reply pkg/password/password.go:41]: resource reconciliation would be one good option, but that one is complicated.

**Comment by pavansokkenagaraj** [comment]
```bash
➜  curl --location --request PUT 'https://kotsadm-pavansokkenagaraj.replicated.okteto.dev/api/v1/password/change' \
--header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzZXNzaW9uSWQiOiIyOElnSmlOQlhyQVlRcEVkdkVvNEM1M3dMcHQifQ.z-lJxZ0Y_jE9TJ_xdV-P-a4BA41PcwyejZmobUSwBFs' \
--header 'Content-Type: application/json' \
--data-raw '{
	"current_password": "passowrds",
    "new_password": "passwords"
}'
```
request
```json
{
    "current_password": "passwords",
    "new_password": "password"
}

```
response: 400 BadRequest
```json
{
    "error": "current password does not match",
    "success": false
}
```
request
```json
{
    "current_password": "password",
    "new_password": "test"
}
```
response: 400 BadRequest
```json
{
    "error": "new password must be at least 6 characters",
    "success": false
}
```
request
```json
{
    "current_password": "password",
    "new_password": "password"
}

```
response: 400 BadRequest
```json
{
    "error": "new password should be different from current password",
    "success": false
}
```
request
```json
{
    "current_password": "password",
    "new_password": "kotsadm"
}

```
response: 200 Ok
```json
{
    "success": true
}
```
